src/System.Runtime.Extensions/src/System/Runtime/CompilerServices/SwitchExpressionException.cs

```diff
@@ -0,0 +1,41 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Serialization;
+
+namespace System.Runtime.CompilerServices
+{
+    /// <summary>
+    /// Indicates that a switch expression that was non-exhaustive failed to match its input
+    /// at runtime, e.g. in the C# 8 expression <code>3 switch { 4 => 5 }</code>.
+    /// The exception optionally contains an object representing the unmatched value.
+    /// </summary>
+    [Serializable]
+    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    public sealed class SwitchExpressionException : InvalidOperationException//, ISerializable
+    {
+        public SwitchExpressionException()
+            : base(SR.Arg_SwitchExpressionException) { }
+
+        public SwitchExpressionException(object unmatchedValue)
+            : base()
+        {
+            UnmatchedValue = unmatchedValue;
+        }
+
+        private SwitchExpressionException(SerializationInfo info, StreamingContext context) 
+            : base(info, context)
+        {
+            info.AddValue(nameof(UnmatchedValue), UnmatchedValue, typeof(object));
+        }
+
+        public object UnmatchedValue { get; }
+
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+           base.GetObjectData(info, context);
+            info.AddValue(nameof(UnmatchedValue), UnmatchedValue, typeof(object));
```

Comment:
What if the type of UnmatchedValue isn't serializable?

Reply:
The purpose of exception serialization in .NET Core is basically telemetry (viz Azure Backup that requested we bring it back). So what I suggest is attempting to serialize, and if there's a SerializationException, create some kind of potentially useful string (using ToString and possibly ellipses?) and stuff that in there instead.

Reply:
>So what I suggest is attempting to serialize

We probably don't need to actually attempt to serialize, but could instead check obj.GetType().IsSerializable.

Regardless, your suggestion for what to do if it's false sounds reasonable: ToString() it and treat that as the UnmatchedValue.

Reply:
What is the pattern to use? i.e. where should I be checking for UnmatchedValue.GetType().IsSerializable?

Reply:
@maryamariyan Dan's suggestion looks fine to me.

Reply:
I agree with @danmosemsft. Given that the value is mostly used during debugging, it's a best effort across serialization boundaries.

Reply:
I do not think it is worth it to have one-off creative fix for this one exception. The `IsSerializable` check is still not going to be enough, e.g. if the matched item is Dictionary containing non-serializable items (https://github.com/dotnet/corefx/issues/35041#issuecomment-459978550
). It is better to be consistently broken.

Reply:
On the other end of the spectrum, the matched item can be a really big serializable object graph. You can end up with megabytes of serializable payload by accident.

If we think about the exception binary serialization should be optimized for telemetry, we may want to *always* convert the object to string for binary serialization to have a better chance for a reasonably sized payload.

On the other end of the spectrum, the matched item can be a really big serializable object graph. You can end up with megabytes of serializable payload by accident.

If we think about the exception binary serialization should be optimized for telemetry, we may want to *always* convert the object to string to have a better chance for a reasonably sized payload.

The binary serialization of exception stack trace is handled in similar way. The stacktrace is converted to string for binary serialization because of the elements that it is constructed are not binary-serialization friendly.

Reply:
In the worst case ToString at least gives me the type name. Im thinking of how many  times had to diagnose something with only the exception details in a log or bug  report. That has made me quite biased towards including any potentially useful information. @stephentoub what are your thoughts? We seem to be having the same discussion in the linked issue 🙂

Reply:
@arunjvs any thoughts, I believe you were or first customer for exception serialization..? See #35041 

Always ToString might work although it could be worse if ToString is not usefully implemented on the type.

Reply:
I am all for having the right level of detail `ToString`.

I just do not think we should spending any cycles on trying to make binary serialization better in creative ways. Binary serialization is a dead end.

Reply:
@maryamariyan please go ahead and submit without the IsSerializable check, and we can have this discussion in #35041